### ISSUE TITLE
Discuss ways to merge separately mapped sidewalks and bike lanes

### ISSUE THREAD

**Issue by tordans** [a-b-street/osm2streets#4]
This conversation started at https://github.com/a-b-street/abstreet/discussions/789 (search term "_separate_" and "_multiple OSM ways_")

## The main question is:

> …how will the library figure out if separate ways are geometrically located to the left or right of the main road?

## Preprocess data

I think we should continue the search for a solution for this. AFAIK there is no way to solve this in JavaScipt/Node. But it is possible to solve it in QGIS and likely the underlying Python libraries.

So it should be possible to preprocess the data in a way, that osm2lanes can then work with.

I think we should check if this could be solved (in a performant way) with Osmium or osm2pgsql or Overpass, … – I don't know enough about those tools ATM.

I assume, one of those tools would be able to merge the separately mapped ways to the central way. This could be done just by proximity, but I assume that would be error prone. But there are tags, that we can use, to reduce the error rate:

## Sidewalks

- `highway=footway + footway=sidewalk` on the separately mapped way is a strong indicator, that the way belongs to the main road

- `sidewalk=separate`/`sidewalk:(left|right|both)=separate` on the main road are a strong indicator, that there is a sidewalk to look for nearby

## Cycleways

- `is_sidewalk=yes` on the separately mapped way is a strong indicator, that the way belongs to the main road. This tag is still experimental, though, but could help a lot. We are experimenting with it to improve routing for footways and cycleways, [more on this wiki page footway mapping (GER)](https://wiki.openstreetmap.org/wiki/Berlin/Verkehrswende/Gehwege#Gehwege.2FB.C3.BCrgersteige) and [this wiki page on cycleways mapping (GER)](https://wiki.openstreetmap.org/wiki/Berlin/Verkehrswende/Radwege) The additional experimental tags `is_sidepath:of=(secondary|residential|<HighwayValueOfMainRoad)` and `is_sidepath:of:name=<NameValueOfMainRoad>` might also be useful, but are likely better automatically processed by the script.

- `cycleway=separate`/`cycleway:(left|right|both)=(separate|optional_sidepath)` on the main road are a strong indicator, that there is a cycleway to look for nearby ([Germany wiki page](https://wiki.openstreetmap.org/wiki/Berlin/Verkehrswende/Radwege))

**Comment by BudgieInWA** [comment]
Now that we are in osm2streets, lets look at ideas for achieving snapping.

One perspective is that of the planar *cough*z-index*cough* graph. When snapping occurs, we are removing a little region in the graph. If we look at the geometry of the regions in the graph, we might be able to recognise medians and verges and thin barriers and stuff, by their characteristics.

**Comment by dabreegster** [comment]
That's a neat idea. I've done some work in tracing polygons "around the block" that could be lifted to the `RawMap` layer, and it usually picks up the spaces in between dual carriageways at least:
![Screenshot from 2022-05-18 09-16-33](https://user-images.githubusercontent.com/1664407/168991843-b8860e72-2c9f-4025-88fa-2cf2c36cd130.png)
You could play with this by going to http://play.abstreet.org/0.3.21/abstreet.html?--dev&system/us/seattle/maps/phinney.bin&--cam=17.89/47.68380/-122.34385, pressing **ctrl+D** for debug mode, then **ctrl+B** for the blockfinder. This process breaks down when the separate ways are so close together that their inferred width causes the two polygons to overlap, though.

**Comment by dabreegster** [comment]
I made recent progress in #61. Projecting perpendicular lines is tricky business for many reasons mentioned above. I tried a new graph-based approach. The parallel cycleway will be connected to the main road at some point -- probably when it crosses side roads. So this transformation uses that, but still uses geometry to figure out left/right.

I'll keep iterating on this particular transformation to make it work more robustly. Then I don't see any reason why we can't also try to apply it to sidewalks...
